### ISSUE TITLE
Implement new rules for conditional components

### ISSUE THREAD

**Issue by casella** [OpenModelica/OpenModelica#12888]
The MAP-Lang group just approved new rules for Modelica 3.7, that relax the very tight restrictions on the use of conditional components, see modelica/ModelicaSpecification#3556. We should start thinking how to implement them in OMC.

@perost commented on Slack:
> But I guess the new rules for conditional components are basically that they can be used like normal components unless they are deleted, in which case they are have restriction where they can be used. Removing the current restriction in the compiler should be trivial, but making sure that deleted components  that are used incorrectly don't break things might take a bit of work.

I would suggest to start adding a flag to remove the restrictions, so we can test stuff. Then, we can take care of the other part of the work. @perost do you agree?

**Comment by perost** [comment]
> I would suggest to start adding a flag to remove the restrictions, so we can test stuff. Then, we can take care of the other part of the work. @perost do you agree?

I made #12889 which disables the check when using `--std=experimental`. But I forgot that we actually just issue a warning for it (unless you use `--strict`), so using the flag isn't strictly necessary.

**Comment by casella** [comment]
I agree it's not strictly necessary, but one would also like to avoid cluttering the compiler output with unnecessary warning, which may cause one to miss the relevant ones.

One more thing, this change was already approved by the language group, so I would classify it as 3.7, as it will certainly belong to the next release. What do you think?

**Comment by perost** [comment]
> One more thing, this change was already approved by the language group, so I would classify it as 3.7, as it will certainly belong to the next release. What do you think?

Yes, and I first added 3.7 to the `--std` flag. But the compiler currently defaults to the latest version if no MSL is loaded, which would've defeated the purpose of putting this behind a flag. I just used the `experimental` version for now instead, since it's meant for these kinds of things, but I will change it to 3.7 once we actually implement the new rules.

**Comment by casella** [comment]
@perost is this issue settled? I compiled
```modelica
model M
  parameter Boolean b = true;
  Real x if b;
equation
  if b then
    x = 2;
  end if;
end M;
```
with the latest nightly and OMC didn't issue any warning.

**Comment by abuntrock** [comment]
I realized this as well with one of our more complex models. Would be great if this is not by accident and can be made available with the 1.26.0 version

**Comment by perost** [comment]
The warning was removed due to #13923. I wouldn't say that this is entirely resolved since we don't check that components are used correctly if they are disabled, but models that are legal according to the new rules should work fine.

**Comment by casella** [comment]
> I realized this as well with one of our more complex models. Would be great if this is not by accident

Absolutely not 😃 

> and can be made available with the 1.26.0 version

It will. 

In fact, this is a game changer, allows do do many interesting things. For example, in the BaseProperties of Modelica.Media we had to choose which variables should be always computed by the fluid equations of state, and then all the extra properties (like entropy, transport properties, partial derivatives) are computed on demand by function calls. This works, but prevents the use of Modelica.Media in context where you want the model to be fully equation-based, e.g. doing optimal control by collocation. Now, we can do it properly with conditional components.

**Comment by casella** [comment]
> The warning was removed due to [#13923](https://github.com/OpenModelica/OpenModelica/issues/13923). I wouldn't say that this is entirely resolved since we don't check that components are used correctly if they are disabled, but models that are legal according to the new rules should work fine.

OK. I will close this ticket because we have some usable functionality in place, so it's good to mention that in the release notes.

#14646 follows up with the remaining open issues.

**Comment by casella** [comment]
PR #14647 updates the User's Guide accordingly.

**Comment by abuntrock** [comment]
We are really looking forward to it. It allows us to get rid of a lot of workarounds in our library. Thanks a lot for implementing that new specification feature this fast.

**Comment by casella** [comment]
>Thanks a lot for implementing that new specification feature this fast.

I was among the ones that pushed for it 😄
